In [1]:
import numpy as np
import matplotlib.pyplot as plt
import scipy

In [2]:
np.random.seed(42)

Part (a)
---

We perform a Wald test for the equality of the means $\mu_1$ and $\mu_2$. The null hypothesis is $H_0:\mu_1=\mu_2$. The difference $\theta=\mu_1-\mu_2$. So the null hypothesis is $H_0:\theta=0$. We compute the plug-in estimator $\hat{\theta}$ of the difference $\theta$.

In [3]:
Twain = np.array([.225, .262, .217, .240, .230, .229, .235, .217])

In [4]:
Snodgrass = np.array([.209, .205, .196, .210, .202, .207, .224, .223, .220, .201])

In [5]:
mu_1_hat = np.mean(Twain)

In [6]:
mu_2_hat = np.mean(Snodgrass)

In [7]:
theta_hat = mu_1_hat - mu_2_hat

In [8]:
print('Mean proportion of 3-letter words in Twain essays: {}'.format(round(mu_1_hat, 3)))
print('Mean proportion of 3-letter words in Snodgrass essays: {}'.format(round(mu_2_hat, 3)))
print('Plug-in estimator for difference: {}'.format(round(theta_hat, 3)))

Mean proportion of 3-letter words in Twain essays: 0.232
Mean proportion of 3-letter words in Snodgrass essays: 0.21
Plug-in estimator for difference: 0.022


We estimate the standard error of $\hat{\theta}$ by $\widehat{\text{se}}=\sqrt{\widehat{\text{se}}(\mu_1)^2 + \widehat{\text{se}}(\mu_2)^2}$

In [9]:
se_1_hat = np.std(Twain)

In [10]:
se_2_hat = np.std(Snodgrass)

In [11]:
se_hat = np.sqrt(se_1_hat**2 + se_2_hat**2)

In [12]:
w = theta_hat / se_hat

In [13]:
print('Estimated standard error of mean proportions from Twain: {}'.format(round(se_1_hat, 3)))
print('Estimated standard error of mean proportions from Snodgrass: {}'.format(round(se_2_hat, 3)))
print('Estimated standard error of difference: {}'.format(round(se_hat, 3)))
print('Observed Wald statistic: {}'.format(round(w, 3)))

Estimated standard error of mean proportions from Twain: 0.014
Estimated standard error of mean proportions from Snodgrass: 0.009
Estimated standard error of difference: 0.016
Observed Wald statistic: 1.35


By Theorem 10.13 the p-value for the Wald test is $2\Phi(-w)$. A 95% confidence interval for $\theta$ is $(\hat{\theta} - 2\widehat{\text{se}}, \hat{\theta} + 2\widehat{\text{se}})$.

In [14]:
from scipy.stats import norm

In [15]:
wald_p_val = 2*norm.cdf(-w)

In [16]:
print('Estimated p-value: {}'.format(round(wald_p_val, 3)))
print('Estimated 95% confidence interval: ({}, {})'.format(
    round(theta_hat - 2*se_hat, 3), round(theta_hat + 2*se_hat, 3)))

Estimated p-value: 0.177
Estimated 95% confidence interval: (-0.011, 0.055)


Thus, using a size threshold of 0.05, we do not reject the null hypothesis that $\mu_1 - \mu_2 = \theta = 0$ using the Wald test. I.e. we do not conclude that Twain and Snodgrass are different authors.

Part (b)
--

Now we use the permutation test to test the null hypothesis that $\mu_1-\mu_2=\theta=0$. Thus, we will consider the tuple $(X_1,\ldots,X_m,Y_1,\ldots,Y_n)$ where $X_i$ are the proportions of 3-letter words in the Twain essays and $Y_i$ are the proportions of 3-letter words in the Snodgrass essays. We will consider $t_{\text{obs}} = \overline{X_m} - \overline{Y_n}$. We will then permute $(X_1,\ldots,X_m,Y_1,\ldots,Y_n)$ randomly to form $(X_1',\ldots,X_m',Y_1',\ldots,Y_n')$ some large number $B$ of times. We will compute the difference $\overline{X_m'} - \overline{Y_n'}$ of means. The p-value will be estimated as the mean of $I(\overline{X_m'} - \overline{Y_n'} > t_{\text{obs}})$ over all these permutations.

In [17]:
t_obs = theta_hat

In [18]:
m = len(Twain)

In [19]:
n = len(Snodgrass)

In [20]:
B = int(1e6)

In [21]:
T_boot = np.empty(B)

In [22]:
full_ds = np.concatenate((Twain, Snodgrass))

In [23]:
for i in range(B):
    perm = np.random.permutation(full_ds)
    mu = np.mean(perm[:m])
    nu = np.mean(perm[m:])
    T_boot[i] = 1 if nu-mu > t_obs else 0

In [24]:
perm_p_val = np.mean(T_boot)

In [25]:
print('p-value from permutation test: {}'.format(round(perm_p_val, 6)))

p-value from permutation test: 0.00026


Since the estimated p-value $0.00026 < 0.05$, using the permutation test, we reject the null hypothesis that $\mu_1=\mu_2$. I.e. we conclude that Twain and Snodgrass are different authors.